modules/drivers/bigquery-cloud-sdk/src/metabase/driver/bigquery_cloud_sdk.clj

```
@@ -155,6 +159,60 @@
+    (when-not (= (class v) Object)
+      (parser v))))
+
+(defn- extract-fingerprint [field-idxs parsers ^FieldValueList values]
+  (map (fn [^Integer idx parser]
+         (parse-field-value (.get values idx) parser))
+       field-idxs parsers))
+
+(defn- sample-table
+  "Process a sample of rows of fields corresponding to the Metabase fields
+  `fields` from the BigQuery table `bq-table` using the query result reducing
+  function `rff`.
+
+  `.getSchema` returns nil if called on the result of `.list`, so we have to
+  match fields by position. Here it is assumed that :database_position in
+  `fields` represents the positions of the columns in the BigQuery table and
+  that `.list` returns the fields in that order. The first assumption could be
+  lifted by matching the names in `fields` to the names in the table schema."
+  [^Table bq-table fields rff]
+  (let [field-idxs  (mapv :database_position fields)
+        all-parsers (get-field-parsers (.. bq-table getDefinition getSchema))
+        parsers     (mapv all-parsers field-idxs)
+        rows        (.list bq-table (u/varargs BigQuery$TableDataListOption))]
+    (transduce (comp (take metadata-queries/max-sample-rows)
+                     (map (partial extract-fingerprint field-idxs parsers)))
+               ;; Instead of passing on fields, we could recalculate the
+               ;; metadata from the schema, but that probably makes no
+               ;; difference and currently the metadata is ignored anyway.
+               (rff {:cols fields})
+               (-> rows .iterateAll .iterator iterator-seq))))
+
+(defmethod driver.workarounds/table-rows-sample :bigquery-cloud-sdk
+  [driver {table-name :name, dataset-id :schema :as table} fields rff opts]
+  (let [database (table/database table)
+        bq-table (get-table database dataset-id table-name)]
+    (if (#{TableDefinition$Type/MATERIALIZED_VIEW TableDefinition$Type/VIEW} (.. bq-table getDefinition getType))
```

From the [possible types](https://cloud.google.com/java/docs/reference/google-cloud-bigquery/latest/com.google.cloud.bigquery.TableDefinition.Type) these are the ones that I could test and know that they are not supported by `.list`. `MODEL` doesn't seem to show up as table. I couldn't easily test `EXTERNAL` and `SNAPSHOT`.

---

> As it stands now, they would result in an exception if they showed up as tables and we tried to fingerprint them.

Do you mean `EXTERNAL` and `SNAPSHOT` don't work with the fingerprint code from _before_ this change? (The default `table-rows-sample` query)? Maybe it's worth logging an error message or something to that effect so if someone actually runs in to it in real life this won't be too confusing to debug

---

Sorry, I didn't formulate that well. What I really mean is that _if_ `EXTERNAL` and `SNAPSHOT` are not supported by `.list` (I think this is likely), the current code would fail fast with an exception. I expect this to be easy to debug.

As an alternative, we can add them to the list of types known not to be supported by `.list`, in which case we would preserve the behavior we have today. @camsaul, which way do you prefer?